plugins/com.yakindu.solidity/src/com/yakindu/solidity/typesystem/BuiltInDeclarationsFactory.java

```diff
@@ -0,0 +1,87 @@
+/**
+ * Copyright (c) 2018 committers of YAKINDU and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ * 
+ * Contributors:
+ * 	Andreas Muelder - Itemis AG - initial API and implementation
+ * 	Karsten Thoms   - Itemis AG - initial API and implementation
+ * 	Florian Antony  - Itemis AG - initial API and implementation
+ * 	committers of YAKINDU 
+ * 
+ */
+package com.yakindu.solidity.typesystem;
+
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.emf.ecore.resource.Resource;
+import org.eclipse.xtext.EcoreUtil2;
+import org.yakindu.base.types.Operation;
+import org.yakindu.base.types.TypesFactory;
+import org.yakindu.base.types.typesystem.ITypeSystem;
+
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import com.yakindu.solidity.solidity.PragmaDirective;
+import com.yakindu.solidity.solidity.SolidityFactory;
+import com.yakindu.solidity.solidity.SourceUnit;
+
+/**
+ * @author Florian Antony - Initial contribution and API
+ */
+@Singleton
+public class BuiltInDeclarationsFactory {

```

Comment:
Maybe we should change the naming to BuiltInDeclarationsProvider and use Guice Provider / assistant inject for the version lookup?

Reply:
From what I understand the solution using [AssistedInject](https://github.com/google/guice/wiki/AssistedInject) will not immediately solve the problem.
In fact we just add an additional indirection without any benefit. The point is that we currently are only able to decide which pragma is relevant within the ScopeProviders. But since the we already have initialized the language infrastructure, we have to calculate the pragma based on the context. In my eyes the only clean solution would be to instantiate the entire language infrastructure (scoping, quick-fixes, builtInDeclarations, type system) based on the pragma. 
Then I could say: inject language infrastructure for 4.7.11. I will create a new issue to further discuss this. 
Right now I think we need to dynamically create BuiltInDeclarations on the fly -> AssistedInject is of no use :/